%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Please note that whilst this template provides a
% preview of the typeset manuscript for submission, it
% will not necessarily be the final publication layout.
%
% letterpaper/a4paper: US/UK paper size toggle
% num-refs/alpha-refs: numeric/author-year citation and bibliography toggle

\documentclass[a4paper,num-refs,gigabyte]{oup-contemporary}

%%% Journal toggle; only specific options recognized.
%%% (Only "gigabyte", "gigascience" and "general" are implemented now. Support for other journals is planned.)
\journal{gigabyte}

\usepackage{graphicx}
\usepackage{siunitx}

%%% Flushend: You can add this package to automatically balance the final page, but if things go awry (e.g. section contents appearing out-of-order or entire blocks or paragraphs are coloured), remove it!
% \usepackage{flushend}

\title{SMARTER-database: a tool to explore genomic diversity in small ruminants}

%%% Use the \authfn to add symbols for additional footnotes, if any. 1 is reserved for correspondence emails; then continuing with 2 etc for contributions.
\author[1,\authfn{1}]{Paolo Cozzi}
\author[1]{Arianna Manunza}
\author[1]{Johanna Ramirez-Diaz}
\author[2,3]{Valentina Tsartsianidou}
\author[2,3]{Konstantinos Gkagkavouzis}
\author[4]{Pablo Peraza}
\author[5]{{Anna Maria} Johansson}
\author[6]{{Juan José} Arranz}
\author[7]{Fernando Freire}
\author[8]{Szilvia Kusza}
\author[1]{Filippo Biscarini}
\author[9]{Lucy Peters}
\author[9]{Gwenola Tosser-Klopp}
\author[4]{Gabriel Ciappesoni}
\author[2,3]{Alexandros Triantafyllidis}
\author[9]{Rachel Rupp}
\author[9]{Bertrand Servin}
\author[1]{Alessandra Stella}

\affil[1]{Institute of Agricultural Biology and Biotechnology, National Research Council, Milano, Italy}
\affil[2]{Department of Genetics, Development \& Molecular Biology, School of Biology, Aristotle University of Thessaloniki, Greece}
\affil[3]{Genomics and Epigenomics Translational Research (GENeTres), Center for Interdisciplinary Research and Innovation (CIRI-AUTH), Greece}
\affil[4]{Sistema Ganadero Extensivo, Instituto Nacional de Investigación Agropecuaria, INIA Las Brujas, Uruguay}
\affil[5]{Department of Animal Breeding and Genetics, Swedish University of Agricultural Sciences, Uppsala, Sweden}
\affil[6]{Departamento de Producción Animal, Facultad de Veterinaria, Universidad de León, León, Spain}
\affil[7]{OVIGEN, Zamora, Spain}
\affil[8]{Centre for Agricultural Genomics and Biotechnology, University of Debrecen, Debrecen, Hungary}
\affil[9]{GenPhySE, Université de Toulouse, INRAE, ENVT, Castanet-Tolosan, France}

%%% Author Notes
\authnote{\authfn{1}paolo.cozzi@ibba.cnr.it}

%%% Paper category
\papercat{Technical Release}

%%% "Short" author for running page header
\runningauthor{Cozzi et al.}

%%% Should only be set by an editor
\jvolume{00}
\jnumber{0}
\jyear{2024}

\begin{document}

\begin{frontmatter}
\maketitle
\begin{abstract}
Underutilized sheep and goat breeds have the ability to adapt to challenging environments due to their genetic composition. Integrating publicly available genomic datasets with new data will facilitate genetic diversity analyses; however, this process is complicated by important data discrepancies, such as outdated assembly versions or different data formats. Here we present the SMARTER-database, a collection of tools and scripts to standardize genomic data and metadata mainly from SNP chips arrays on global small ruminant populations with a focus on reproducibility. SMARTER-database harmonizes genotypes for about 12,000 sheep and 6,000 goats to a uniform coding and assembly version. Users can access the genotype data via FTP and interact with the metadata through a web interface or programmatically using their custom scripts, enabling efficient filtering and selection of samples. These tools will empower researchers to focus on the crucial aspects of adaptation and contribute to livestock sustainability, leveraging the rich dataset provided by the SMARTER-database.
\section{Availability \& Implementation}
The code is available as open source software under the MIT license at \url{https://github.com/cnr-ibba/SMARTER-database}.
\end{abstract}

\begin{keywords}
sheep; goat; genotypes; adaptation; REST-API; standardization; reproducibility; database
\end{keywords}
\end{frontmatter}

\section{Statement of Need}

\subsection{Background}

The presence of small ruminant populations is crucial to the socio-economic prosperity of human settlements, particularly in European marginal regions. In these areas, sheep and goat breeds that are not fully utilized have the potential to significantly increase the profitability of small ruminant farming. The reason for their value is their distinctive and often unusual genetic composition (e.g. \citep{Stella2018,Kijas2012}), which renders them an exceptionally valuable resource for adapting to challenging environments, withstanding harsh farming conditions, combating biotic and abiotic stressors, and producing high-quality animal-derived products. In this context, the SMARTER project\citep{SmarterProjectEU} developed innovative strategies to improve resilience and efficiency related traits of sheep and goats in diverse environments.
Here we present the SMARTER-database, a collection of tools and scripts to gather, standardize and provide the scientific community with a comprehensive dataset of genomic data and metadata information on worldwide small ruminant populations. Existing datasets were scouted from public repositories and complemented with newly produced data within the context of the SMARTER project\citep{SmarterProjectEU}. Our system provides a single entry point and standardization tools to explore genetic diversity and demography of goat and sheep breeds, and to understand the genetic basis of resilience and adaptation, especially in under-utilized breeds.

\subsection{Data composition}

SMARTER-database is mainly composed of two types of data: i. genotype data, derived from low/high density genome chips and Whole Genome Sequencing (WGS); ii. phenotype data, including a wide range of information such as GPS data relative to sampled populations, morphological description of animals and other production measurements.
The objective of our work was to integrate numerous publicly available and newly generated datasets into a single entry point.
For genotype datasets, the process of data integration is complicated by the fact that several public datasets were generated years ago and refer to outdated genome assembly versions, using different variant names assigned by different SNP array manufacturers, or encoding the same information in different ways.
Phenotype data is even more heterogeneous. Missing information is hard to retrieve since animals may no longer be available. Phenotypic data collection often aligns closely with specific experimental objectives, yet descriptions of such data seldom adhere to validated ontological frameworks (e.g. PATO - the Phenotype And Trait Ontology\citep{PATO}). This lack of standardization complicates cross-study comparisons, hindering robust cross-referencing and comprehensive analytical endeavors. Since the focus of the SMARTER project was on adaptation, a minimal set of attributes was identified for newly collected data: country of origin, breed name, internal IDs used in the genotype file, GPS coordinates and the main purpose of the breed when available (dairy, meat, wool).  We however encourage submission of any type of information relevant to resilience, efficiency and adaptation potential.

Since most of the data are derived from SNP chip arrays, the SMARTER database is modeled around this type of information. WGS data, which represents a small fraction of the total data, is filtered and merged relying on the SNPs present in the SNP arrays. To integrate genotype datasets generated at different times and with different technologies, all files must first be converted to the same format. The majority of data submissions were formatted in PLINK\citep{Purcell2007}, followed by Illumina and Affymetrix formats. Most software packages lack support for proprietary file formats, such as Illumina’s row files or Affymetrix’s cell data files. Consequently, prior to merging, these proprietary files must undergo conversion into universally accepted formats.

Furthermore, it is imperative to ensure consistency in the reference genome assembly across all datasets. Discrepancies in assembly versions can lead to variations in SNP positions. Even within the same assembly, SNPs may exhibit divergent locations if the mapping procedures between two genotyping array types are not identical\citep{Fadista2012,Gershoni2022}. Reliance solely on SNP names was found to be inadequate, given that SNPs from different manufacturers frequently possess unique identifiers and may lack comprehensive information required for unambiguous SNP identification across diverse datasets.

The encoding of genotypes presents another compatibility challenge, as two different formats may be employed to convey identical SNP information. SNPs are identified by aligning a SNP probe (comprising a brief DNA sequence around the SNP) to the reference genome, which may be on either the forward or reverse strand across various genome assemblies. As a result, the same SNPs can be represented by different bases: for example, an [A/G] polymorphism becomes [T/C], if the probe matches on different strands. In such cases, the FAO Guidelines for Genomic characterization of animal genetic resources\citep{Ajmone23} recommend to reverse the SNP before merging the datasets, and to exclude the SNP when the first base is complementary to the second (ie [A/T] - [C/G]), since in these cases it is not possible to verify if the SNP is on the same strand on both assemblies without additional information.
Besides, in the A/B format the letters A and B don’t represent the real genotype but the first and second letters of the SNP recorded in the manifest file: in such cases it’s mandatory to get access to the information in the same manifest file used in the genotyping process.
To avoid problems caused by probe alignment and ensure consistent SNP representation regardless of genome version, the company Illumina has proposed a TOP/BOTTOM format\citep{IlluminaTOP}, which relies on the SNP probe sequences themselves, rather than on probe alignments as in the forward/reverse convention. This solution is not widely adopted: for example, to submit data to EBI-EVA\citep{EVASubmit}, one must provide SNPs in forward orientation with respect to the reference genome (the reference allele needs to match the reference genome in the same position\citep{AssemblyChecker}).
In 2015, we responded to the need to standardize SNP data\citep{Nicolazzi2015a} by developing SNPchiMp\citep{Nicolazzi2015b} and a series of tools to work with different data sources and to convert data in the same format\citep{Nicolazzi2016}. In SMARTER, we followed up on the same concept: development of tools to convert genomic data to a reference format (i.e. Illumina TOP), and then giving the possibility to convert the data according to user needs, such as publishing data to public repositories or before merging data with other datasets encoded in a different way. An example of genotype conversion for \emph{unambiguous} and \emph{ambiguous} SNP is presented in Table \ref{tab:atogSNP} and Table \ref{tab:ctogSNP} respectively. During the genotype data import, genotypes are converted to Illumina TOP, in order to refer to the same SNP coding across samples from different datasets. It is important to note that after the Illumina TOP conversion, the only available genotypes will be [A/T], [C/G], [A/G], or [A/C], as all possible allele combinations can be converted to one of these four.

\begin{table*}[bt!]
\caption{Genotype conversion of DU186191\_327.1 (A/G) SNP for four different samples in the SMARTER database. The source version is the assembly version of the received samples, and source coding and genotype are the inferred coding and the received genotype respectively. The TOP genotype is the final genotype present in database according to Illumina TOP\citep{IlluminaTOP} convention}
\label{tab:atogSNP}
\begin{center}
\begin{tabular}{l l l l l}
\toprule
smarter\_id & source version & source coding & source genotype & TOP genotype \\
\midrule
UYOA-CRR-000003890 & Oar\_v4.0 & forward & T C & A G \\
UYOA-CRL-000000382 & Oar\_v3.1 & A/B & A B & A G \\
NAOA-ADP-000001020 & Oar\_v3.1 & top & G A & G A \\
GROA-CHI-000004137 & Oar\_v3.1 & forward & T T & A A \\
\bottomrule
\end{tabular}
\end{center}
\end{table*}

\begin{table*}[bt!]
\caption{Genotype conversion of OAR1\_103790218.1 C/G SNP for four different samples in the SMARTER database. The source version is the assembly version of the received samples, and source coding and genotype are the inferred coding and the received genotype respectively. The TOP genotype is the final genotype present in database according to Illumina TOP\citep{IlluminaTOP} convention}
\label{tab:ctogSNP}
\begin{center}
\begin{tabular}{l l l l l}
\toprule
smarter\_id & source version & source coding & source genotype & TOP genotype \\
\midrule
UYOA-CRR-000003890 & Oar\_v4.0 & forward & G G & G G \\
UYOA-CRL-000000382 & Oar\_v3.1 & A/B & B B & G G \\
NAOA-ADP-000001020 & Oar\_v3.1 & top & G G & G G \\
GROA-CHI-000004137 & Oar\_v3.1 & forward & C C & G G \\
\bottomrule
\end{tabular}
\end{center}
\end{table*}

\subsection{Collecting phenotypes and other metadata}

Besides genotype data, other types of data can also be useful when processing and analyzing genotypes in the context of adaptation and genomic selection. Genotype file formats are not suitable for the inclusion of data like geographical coordinates, or morphological description of animals. Even more advanced standards like Variant Calling Format (VCF) provide an INFO field supporting user defined data\citep{Danecek2011}, accessing this kind of information is inefficient  because this field cannot be  indexed.  Preferably, metadata information should be stored in a dedicated resource where it can easily be queried. In the context of the SMARTER project, a schemaless database was more desirable since metadata don't follow a standard format. A minimal set of requirements should be provided in order to make useful queries: for instance, breed, country and GPS locations are required in order to retrieve data relevant for adaptation analyses.

\section{Methods}

\subsection{The SMARTER-database project}

In order to standardize data and merge genotypes from different datasets, we collected information from SNP chip manufacturers both by accessing publicly available manifest files and by directly contacting the manufacturer in case of custom manifest files. Data stored in the SNPchiMp database\citep{Nicolazzi2015b} was also collected as a unique source of information for SNP positions and code conversion. In addition, information from public databases like dbSNP\citep{Sherry1999} and EVA\citep{Cezard2022} is integrated in order to provide external reference IDs to SNPs.
To store metadata information, we decided  to employ a MongoDB\citep{MongoDBURL} database. This type of database, characterized by its schemaless nature and support for spatial queries, enables the flexible modeling of data with the capability to dynamically add or remove attributes as needed. In addition to sample information, the database is designed to also accommodate variant information, thereby facilitating genotype conversion through reliance on the database. The variant class model defined with the help of the MongoEngine\citep{MongoEngine} library, a Python-based Object-Document Mapper (ODM)\citep{ODBvsORM} that facilitates working with MongoDB in an object-oriented way, is presented in Figure \ref{fig:variant_class_model}: a VariantSpecie abstract class defines all the attributes that can be referred to the same SNP, like the different names present in different chip manufacturers or external accession IDs. Multiple Locations are then embedded in the same document in order to track different positions across different assemblies with their specific allele coding. Finally, the VariantSpecie abstract class is extended by the VariantGoat and VariantSheep classes, which serve as the final Object Document Mappers\citep{ODBvsORM} for each sheep and goat variant in the database. All information related to the same SNP is present in the same document allowing to track the same SNPs even when named differently by each technology. Moreover, it is possible to identify which SNPs are in common between different datasets and to query data relying on manufacturer's variant name, rsID and genomic locations. Finally, the MongoEngine\citep{MongoEngine} implementation defined some accessory methods which have no direct representation in MongoDB itself but can be exploited to determine the genotype coding or the genotype conversion of a provided SNP according to the assembly information present in the database. More information about the variant model is in the SMARTER-database documentation\citep{SMARTERdocs}.

Although VCF\citep{Danecek2011} file format was proposed as a standard for the distribution of genotype information, we choose to first merge all genotypes we received in a unique PLINK binary file, one for goats and one for sheep. Proprietary file formats (e.g. illumina row and Affymetrix cell files) were converted to PLINK files and merged with other data using PLINK\citep{Purcell2007} software. The reason for adopting this format is related to its popularity in the research community: many software applications, libraries and pipelines focusing on adaptation or genetic diversity rely on this format. In addition, the much smaller size of the PLINK binary file makes it easy to manage (to subset, to index, and more in general to analyze) data using the PLINK software.

The Illumina TOP/BOTTOM coding convention was selected due to its reliance solely on the probe itself\citep{IlluminaTOP}. Its primary advantage lies in ensuring consistent encoding of SNPs across various assembly versions. Consequently, updating SNP positions to accommodate a new reference assembly merely entails a straightforward adjustment.

Different attributes linked to the same SNP, including varying names used by Affymetrix and Illumina or differences in coding types, are leveraged to produce a unified genotype file. This consolidated file guarantees SNP standardization across different datasets by assigning them uniform names and codes, thereby streamlining analyses across samples from various datasets. Furthermore, sample metadata stored in SMARTER-database helps to identify relevant samples, simplifying their extraction from the comprehensive genotype file relying on the same sample IDs.

The final intersection among all different chip types in sheep will be close to 30K SNPs, including various versions of Illumina 50K chip (which ranges between 50K and 60K SNPs depending on the release), HD chip and public and custom Affymetrix chips. This implies that at least 30K SNPs are theoretically shared across all 12K sheep samples, however all remaining SNPs are reported in the final genotype to support nearly all the 620K sheep SNPs managed by the database. Samples without information on SNPs outside the chip used to generate them will have missing data. The same approach was adopted for goats, however, the number of supported chips is lower resulting in a final intersection of nearly 93\% of the goat supported SNPs.

Data are provided as they are, with no filtering applied to the final dataset: users are required to apply their filters after selecting the data they need, thus avoiding filtering by missingness using samples irrelevant to their analyses. Standard Minor Allele Frequency (MAF) filters applied as a percentage cannot be applied to the entire dataset, as stated by the FAO guidelines\citep{Ajmone23}, since we would lose all the variability associated with local adaptation, and MAF could change depending on different subsets of samples. Additionally, filtering based on assembly position or sex chromosomes could result in a loss of information: for example, when updating an assembly, unmapped SNPs might be mapped in the new assembly, while previously mapped SNPs could become unplaced. Ideally, a custom remapping of Illumina and Affymetrix probes against new genome assemblies could increase the intersection between different chip technologies and this can be added in a future release of the database. Sex chromosomes can also be informative, particularly for users interested in reproduction studies, therefore they aren't removed. Consequently, we believe that this dataset should be presented without any filtering, leaving it to the user to document all the steps needed to produce the final dataset required for their analyses.

\begin{figure*}%[b!]  %% Add a [b!] if you prefer the wide image to be at the bottom of the page
\centering
\includegraphics[width=.7\textwidth]{VariantModel.png}
\caption{A UML diagram of the Variant classes model implemented in Python: the boxes in the diagram show class attributes and methods, with data types like \texttt{str} for string and \texttt{dict} for dictionary. The central \texttt{VariantSpecie} class serves as an \textit{abstract} base class, containing common attributes and methods relevant to different species, such as \textit{SNP IDs}, \textit{chip names}, and \textit{genotype data}. The \texttt{Location} class models assembly-specific information, including \textit{chromosomal positions}, \textit{alleles}, and \textit{strand orientation}, and is embedded within the \texttt{VariantSpecie} class to handle multiple \textit{locations} per variant.
The \texttt{VariantSheep} and \texttt{VariantGoat} classes inherit from \texttt{VariantSpecie}, meaning they extend the base class by specializing it for sheep and goat data, respectively. This inheritance allows for efficient reuse of common functionality while tailoring specific attributes or behaviors for each species. The \texttt{Probeset} class, associated with \texttt{VariantSpecie}, represents specific data types that come from \textit{Affymetrix} chips. It captures metadata linked to SNP probes, such as the \textit{chip name} and \textit{probeset ID}.}
\label{fig:variant_class_model}
\end{figure*}

\subsection{Reproducibility}

The code developed in the SMARTER-database project was enhanced to develop utilities used to maintain the database updated, such as adding new data sources, adding new breeds and samples and convert genotypes in the desired formats in order to produce the final genotype file, as described by our data import guide\citep{DataImportProcess} in the project documentation. The idea behind this implementation was to adhere to the FAIR principles\citep{Wilkinson2016} by providing a reproducible and transparent workflow to create and manage the final dataset.
Our project is based on the Cookiecutter Data Science Project\citep{CookiecutterDataScience}, aiming to standardize data science projects for sharing purposes. It adheres to conventions outlined in the Cookiecutter framework\citep{Cookiecutter}, which includes organizing data into specific folders such as raw data, external data, processing, and final data. Additionally, it includes source folders for importing scripts and libraries, which can be installed as a Python package in a Conda\citep{Anaconda} environment. All software dependencies are managed using Conda - a package and environment management system for Python - and tracked through requirements files. Furthermore, it introduces a database folder not found in the original Cookiecutter template. This folder manages database installation and initialization through Docker\citep{Docker} and docker-compose\citep{DockerCompose}, where Docker is a platform for containerizing applications to ensure they run consistently across different environments.
Raw data undergoes initial exploration using IPython notebooks stored in the notebook folder, aiming to comprehend its structure and potential issues before importing data into the database: efforts are made to infer genotype coding, and to ensure all necessary information is available for genotype conversion and sample addition to the database. Subsequently, the dataset is integrated into the database, with updates to available breeds, if new are added, and the assignment of SMARTER unique and stable IDs to samples, which are used in the resultant genotype files. If metadata is provided, it is appended to the corresponding samples. Currently, the SMARTER database tracks information on approximately 12,000 sheep and 6,000 goats, with nearly 80\% and 60\% of samples possessing GPS coordinates for sheep and goats respectively, as indicated in Table \ref{tab:samples}. In Figure \ref{fig:sheep_samples} and \ref{fig:goat_samples} a graphical representation of samples by country for sheep and goats respectively is reported.
Figure \ref{fig:sheep_technology} illustrates the distribution of 12,000 sheep samples based on the genotype technology employed. Despite variations in genotyping technologies, a subset of SNPs are in common: this enables comparison among samples from different datasets.
All data processing steps, including environment setup, database initialization, and data processing, are handled using GNU Make\citep{Stallman2004} in order to provide simple commands to execute all the steps required to produce the final database. These steps, and the importing scripts executed by these steps, are idempotent, meaning that repeating the same command has no side effects, ensuring consistency in the final dataset. This also means that a new dataset can be added simply by updating the Makefile with the new required steps and then calling the proper make command again. The database follows Semantic Versioning\citep{prestonwerner2013semantic} to track updates and changes effectively. This project is publicly accessible on GitHub\citep{SMARTERDB-GH}, where GitHub workflows automate the process of running Continuous Integration (CI) tests which validate critical steps, such as genotype conversion, to ensure that everything functions correctly after each new code contributions.

\begin{table}
\caption{Amount of samples with GPS coordinates. Foreground samples are produced within the SMARTER project, while background samples are collected from already published datasets}
\label{tab:samples}
\begin{center}
\begin{tabular}{l l l}
\toprule
& Total & With GPS \\
\midrule
Foreground Sheep & 5945 & 79,34\% \\
Background Sheep & 5892 & 82,30\% \\
Foreground Goats & 464 & 87,28\% \\
Background Goats & 5540 & 58,34\% \\
\bottomrule
\end{tabular}
\end{center}
\end{table}

\begin{figure*}
\centering
\includegraphics[width=.95\textwidth]{sheep-by-country.png}
\caption{Geographical distribution of sheep samples in the SMARTER database. Countries are colored based on sample count using the viridis scale, where brighter colors represent higher sample counts.}
\label{fig:sheep_samples}
\end{figure*}

\begin{figure*}
\centering
\includegraphics[width=.95\textwidth]{goat-by-country.png}
\caption{Geographical distribution of goat samples in the SMARTER database. Countries are colored based on sample count using the viridis scale, where brighter colors represent higher sample counts.}
\label{fig:goat_samples}
\end{figure*}

\begin{figure}
\centering
\includegraphics[width=.45\textwidth]{sheep-samples-by-chip-type.png}
\caption{Distribution of sheep samples by genotyping technology}
\label{fig:sheep_technology}
\end{figure}

\subsection{Data access}

Data can be accessed through two distinct channels: the fully processed PLINK binary genotype files for both goat and sheep are accessible via FTPS, while sample information and metadata are retrievable through SMARTER-backend\citep{SMARTERAPI}, a RESTful API interface\citep{REST}. Users are required to identify their desired samples using the API interface and then to extract their genotypes from the global file. This also means that the resource can be programmatically accessed, thereby enhancing reproducibility and facilitating data integration with other resources.

SMARTER-backend is a Python-Flask application built on top of the SMARTER-database MongoDB, offering a specific URL (i.e endpoint) to interact with each object stored in the SMARTER-database. Users can query for SNPs, samples, datasets, and breeds by making HTTP requests to the appropriate endpoint and providing the necessary parameters to obtain information in JSON format (Fig. \ref{fig:backend_url}).
Endpoint parameters documentation is detailed using Swagger and can be accessed via the \texttt{/docs} location of the SMARTER-backend itself. Table \ref{tab:endpoints} provides a list of the available endpoints.
Figure \ref{fig:endpoint} showcases an example of documentation for the \texttt{/sample/goat} endpoint, which enables users to gather information on goat samples. All available parameters are listed along with their descriptions and the supported data types. When the input data type is an array, users can provide the same parameter multiple times. For instance, it is possible to specify multiple breed codes in a single query to obtain all desired samples as illustrated above in Figure \ref{fig:backend_url}.

To facilitate data access for partners primarily focused on data analysis with R rather than retrieving data via REST interfaces, we developed the smarterapi R package\citep{RSMARTERAPI}. This package aims to streamline data retrieval from SMARTER-backend by abstracting the complexities of handling HTTP requests and result pagination. It offers simple R functions that return data as R data frames, utilizing the same parameters described in the backend documentation. This package facilitates URL generation and data retrieval within R.
To assist users, we created an online vignette outlining common operations such as data retrieval and filtering, as well as more advanced tasks like working with variants, conducting spatial queries, and extracting data from raster objects using WorldClim database\citep{Fick2017}. These functionalities are valuable for performing landscape genomic analyses on SMARTER data.

The latest tool provided to the community is SMARTER-frontend\citep{SMARTERfrontend}, a web application developed in Angular\citep{Angular} on the top of SMARTER-backend. This application allows users to browse SMARTER data and gain insights into database contents without the need for additional code or software installation. Regardless of the chosen method for collecting samples (API request, R package, or web application), users must identify their samples of interest and extract the required genotypes using PLINK.

\begin{figure}
\centering
\includegraphics[width=.45\textwidth]{backend_URL.png}
\caption{Structure of SMARTER-backend API URL. The endpoint part specifies the desired data type and the parameters let to filter out results}
\label{fig:backend_url}
\end{figure}

\begin{table*}
\caption{SMARTER-backend available endpoints. The available data types are described in the SMARTER-database online documentation\citep{SMARTERdocs}}
\label{tab:endpoints}
\begin{center}
\begin{tabular}{l l l}
\toprule
Endpoint suffix & Data type & Description \\
\midrule
/auth/login & Users & User authentication (removed after public release)\\
/breeds & Breed & Available breeds as a list \\
/datasets & Dataset & Available datasets as a list \\
/info & SmarterInfo & Information about database status \\
/samples/sheep & SampleSheep & Sheep samples as a list \\
/samples/goat & SampleGoat & Goat samples as a list \\
/samples.geojson/sheep & GeoJSON & Sheep samples in GeoJSON format \\
/samples.geojson/goat & GeoJSON & Goat samples in GeoJSON format \\
/supported-chips & SupportedChip & Supported genotype platforms as a list \\
/variants/sheep/OAR3 & VariantSheep & Supported sheep SNPs in Oar\_v3.1 assembly \\
/variants/sheep/OAR4 & VariantSheep & Supported sheep SNPs in Oar\_v4.0 assembly \\
/variants/goat/ARS1 & VariantGoat & Supported goat SNPs in ARS1.2 assembly \\
/variants/goat/CHI1 & VariantGoat & Supported goat SNPs in CHIR\_1.0 assembly \\
\bottomrule
\end{tabular}
\end{center}
\end{table*}

\begin{figure*}
\centering
\includegraphics[width=.95\textwidth]{endpoints.png}
\caption{Screenshot of the Swagger interface for the SMARTER-backend API. The interface displays available endpoints and parameters, allowing users to interact with and test the API directly. In this example, the \texttt{GET /samples/goat} endpoint is shown, with parameters such as \texttt{breed}, \texttt{breed\_code}, and \texttt{chip\_name}. The \texttt{array[string]} type indicates that multiple string values can be entered, enabling filtering by multiple criteria. Users can click the \textit{'Try it out'} button to input values, execute the query, and view the results with the URL to generate them, making it easy to explore the API's functionality.}
\label{fig:endpoint}
\end{figure*}

\subsection{Conclusions and future developments}
The SMARTER database project provides valuable information on sheep and goat populations around the world. It is an essential tool for researchers, enabling them to generate new insights and offer the possibility to store new genotypes and drive progress in this field. Comprising a suite of scripts, it standardizes genotype data sourced from various methods and origins, resulting in a unified dataset primed for analysis across different assembly versions for both sheep and goats. Data access is granted to users through the SMARTER-backend, using R packages or the web interfaces, while genotypes are available over FTPS. The entire project was developed with the goal of facilitating reproducibility and programmatic data accessibility.
The SMARTER database is open to additional dataset integration: given the schemaless nature of the database and the ability to collect data using the REST API, the most valuable contributions should include metadata, such as GPS coordinates and phenotypes. This type of data will contribute significantly to understanding adaptation and resilience in small ruminants.
We plan to support additional assemblies with the option to collect data in VCF format. This will help the community reuse this data and provide the opportunity to upload genotypes to public archives like EBI-EVA.
All new improvements are tracked as GitHub issues\citep{SMARTERDB-GH-ISSUES}, and when the changes are finalized, they are released in a new dataset version, along with any enhancements to data management. All release changes can be reviewed in the HISTORY.rst\citep{SMARTERDB-GH-HISTORY} file available on GitHub and in the SMARTER-database Read The Docs documentation\citep{SMARTERdocs}.

\section{Availability of source code and requirements}

\begin{itemize}
\item Project name: SMARTER-database
\item Project home page: \url{https://github.com/cnr-ibba/SMARTER-database}
\item Documentation: \url{https://smarter-database.readthedocs.io/en/latest/}
\item Operating system(s): Linux
\item Programming language: Python 3.x
\item Other requirements: docker, docker-compose, anaconda
\item License: MIT
\end{itemize}

\section{Availability of Test Data}

The genotype files supporting the results of this article are available via anonymous FTP at \url{ftp://webserver.ibba.cnr.it} or through \emph{smarterapi} R package (see \url{https://cnr-ibba.github.io/r-smarter-api/articles/smarterapi.html#collect-genotypes}). Metadata and sample information can be accessed via the SMARTER-backend REST API at \url{https://webserver.ibba.cnr.it/smarter-api/} using \emph{smarterapi} R package (\url{https://cnr-ibba.github.io/r-smarter-api/}) and through the database portal at \url{https://webserver.ibba.cnr.it/smarter/}.

\section{Declarations}

\subsection{List of abbreviations}
% If abbreviations are used in the text they should be defined in the text at first use, and a list of abbreviations should be provided in alphabetical order.

\begin{itemize}
\item API: Application Programming Interface
\item CI: Continuous Integration
\item EBI: European Bioinformatics Institute
\item EVA: European Variation Archive
\item FAO: Food and Agriculture Organization
\item FTPS: File Transfer Protocol Secure
\item GNU: GNU's Not Unix
\item GPS: Global Positioning System
\item HTTP: HyperText Transfer Protocol
\item JSON: JavaScript Object Notation
\item MAF: Minor Allele Frequency
\item ODM: Object-Document Mapper
\item PATO: Phenotype And Trait Ontology
\item REST: Representational State Transfer
\item RESTful: REST-compliant systems
\item rsID: Reference SNP cluster ID
\item SNP: Single Nucleotide Polymorphism
\item UML: Unified Modeling Language
\item URL: Uniform Resource Locator
\item VCF: Variant Calling Format
\item WGS: Whole Genome Sequencing
\end{itemize}

\subsection{Ethical Approval}

Not applicable

\subsection{Competing Interests}

The author(s) declare that they have no competing interests

\subsection{Funding}

This project has received funding from the European Union’s Horizon 2020 research and innovation program under grant agreement No 772787 (SMARTER)

\subsection{Author's Contributions}

\begin{itemize}
\item Conceptualization: Alessandra Stella, Bertrand Servin, Gabriel Ciappesoni, Alexandros Triantafyllidis
\item Resources: Valentina Tsartsianidou, Pablo Peraza, {Anna Maria} Johansson, {Juan José} Arranz, Fernando Freire, Szilvia Kusza, Gwenola Tosser-Klopp, Gabriel Ciappesoni, Alexandros Triantafyllidis, Bertrand Servin, Alessandra Stella
\item Methodology: Paolo Cozzi, Bertrand Servin, Alessandra Stella, Alexandros Triantafyllidis, Gwenola Tosser-Klopp
\item Software: Paolo Cozzi, Konstantinos Gkagkavouzis
\item Data curation: Paolo Cozzi, Valentina Tsartsianidou, Konstantinos Gkagkavouzis, Pablo Peraza, {Anna Maria} Johansson, {Juan José} Arranz, Fernando Freire, Szilvia Kusza, Gwenola Tosser-Klopp
\item Formal analysis: Arianna Manunza, Johanna Ramirez-Diaz, Valentina Tsartsianidou, Filippo Biscarini, Lucy Peters
\item Writing - original draft preparation: Paolo Cozzi, Alessandra Stella
\item Writing - review \& editing: Arianna Manunza, Johanna Ramirez-Diaz, Bertrand Servin, Filippo Biscarini, Rachel Rupp
\item Visualization: Paolo Cozzi
\item Supervision: Alessandra Stella, Bertrand Servin
\item Funding acquisition: Alessandra Stella, Bertrand Servin
\item Project administration: Rachel Rupp
\end{itemize}

\section{Acknowledgements}

Not applicable

%% Specify your .bib file name here, without the extension
\bibliography{paper-refs}

\end{document}
